engine/src/execution_graph/logic_controllers/LogicPrimitives.h

```diff
@@ -83,5 +115,314 @@ TableViewPair createEmptyTableViewPair(std::vector<cudf::type_id> column_types,
 
 }  // namespace frame
 
-}  // namespace ral
+namespace cache{
+
+	enum CacheDataType{
+			GPU,
+			CPU,
+			LOCAL_FILE,
+			IO_FILE
+		};
+
+
+
+class CacheData{
+public:
+
+
+	virtual std::unique_ptr<ral::frame::BlazingTable> decache() = 0;
+	virtual unsigned long long sizeInBytes() = 0;
+	virtual ~CacheData(){
+
+	}
+protected:
+	std::vector<std::string> names;
+
+};
+
+class CacheDataLocalFile : public CacheData{
+public:
+  std::unique_ptr<ral::frame::BlazingTable> decache() override;
+
+  CacheDataLocalFile(std::unique_ptr<ral::frame::BlazingTable> table);
+	unsigned long long sizeInBytes() override;
+  virtual ~CacheDataLocalFile(){
+
+  }
+  std::string filePath() const {
+    return filePath_;
+  }
+private:
+  std::string filePath_;
+  //ideally would be a writeable file
+};
+
+
+class CacheMachine{
+public:
+	CacheMachine(unsigned long long gpuMemory, std::vector<unsigned long long> memoryPerCache, std::vector<CacheDataType> cachePolicyTypes);
+	virtual std::unique_ptr<ral::frame::BlazingTable> pullFromCache();
+//  void eject(int partitionNumber);
+//  void reprocess(int partitionNumber);
+	virtual ~CacheMachine();
+	void addToCache(std::unique_ptr<ral::frame::BlazingTable> table);
+//  void reduceCache(int cacheIndex, unsigned long long reductionAmount);
+//  void incrementCache(int cacheIndex, unsigned long long incrementAmount);
+//TODO: figure out how we tell it which cache to prefer
+
+  bool finished();
+	void finish();
+protected:
+	std::mutex cacheMutex;
+	std::vector<std::unique_ptr<std::queue<std::unique_ptr<CacheData> > > > cache;
+  std::queue<std::unique_ptr<ral::frame::BlazingTable> > gpuData;
+//  std::map<int,std::unique_ptr<CacheData> > operatingData;
+
+	std::vector<CacheDataType> cachePolicyTypes;
+  std::vector<unsigned long long> memoryPerCache;
+  std::vector<unsigned long long> usedMemory;
+  bool _finished;
+};
+
+static std::string type_string(cudf::data_type dtype) {
+    using namespace cudf;
+
+    switch (dtype.id()) {
+        case BOOL8: return "BOOL8";
+        case INT8:  return "INT8";
+        case INT16: return "INT16";
+        case INT32: return "INT32";
+        case INT64: return "INT64";
+        case FLOAT32: return "FLOAT32";
+        case FLOAT64: return "FLOAT64";
+        case STRING:  return "STRING";
+        case TIMESTAMP_DAYS: return "TIMESTAMP_DAYS";
+        case TIMESTAMP_SECONDS: return "TIMESTAMP_SECONDS";
+        case TIMESTAMP_MILLISECONDS: return "TIMESTAMP_MILLISECONDS";
+        case TIMESTAMP_MICROSECONDS: return "TIMESTAMP_MICROSECONDS";
+        case TIMESTAMP_NANOSECONDS: return "TIMESTAMP_NANOSECONDS";
+        default: return "Unsupported type_id";
+    }
+}
 
+static void print_blazing_table_view(ral::frame::BlazingTableView table_view, const std::string table_name = ""){
```

Comment:
why is this in here this seems like debugging/testing code

Reply:
yes I removed this dcode